File: element-common/src/main/java/io/github/solusmods/eternalcore/element/impl/data/ElementTagProvider.java
package io.github.solusmods.eternalcore.element.impl.data;

import io.github.solusmods.eternalcore.element.api.Element;
import io.github.solusmods.eternalcore.element.impl.ElementRegistry;
import net.minecraft.core.HolderLookup;
import net.minecraft.data.PackOutput;
import net.minecraft.data.tags.IntrinsicHolderTagsProvider;

import java.util.concurrent.CompletableFuture;

public abstract class ElementTagProvider extends IntrinsicHolderTagsProvider<Element> {
    public ElementTagProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider) {
        super(output, ElementRegistry.KEY, lookupProvider, element -> ElementRegistry.ELEMENTS.getKey(element).orElseThrow());
    }

    public ElementTagProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider, CompletableFuture<TagLookup<Element>> parentProvider) {
        super(output, ElementRegistry.KEY, lookupProvider, parentProvider, element -> ElementRegistry.ELEMENTS.getKey(element).orElseThrow());
    }
}

File: element-common/src/main/kotlin/io/github/solusmods/eternalcore/element/impl/data/ElementTagProvider.kt
package io.github.solusmods.eternalcore.element.impl.data

import io.github.solusmods.eternalcore.element.api.Element
import io.github.solusmods.eternalcore.element.impl.ElementRegistry
import net.minecraft.core.HolderLookup
import net.minecraft.data.PackOutput
import net.minecraft.data.tags.IntrinsicHolderTagsProvider
import java.util.concurrent.CompletableFuture
import java.util.function.Function

abstract class ElementTagProvider : IntrinsicHolderTagsProvider<Element?> {
    constructor(output: PackOutput, lookupProvider: CompletableFuture<HolderLookup.Provider?>?) : super(
        output,
        ElementRegistry.key,
        lookupProvider,
        Function { element: Element? -> ElementRegistry.elements.getKey(element).orElseThrow() })

    constructor(
        output: PackOutput,
        lookupProvider: CompletableFuture<HolderLookup.Provider?>?,
        parentProvider: CompletableFuture<TagLookup<Element?>?>?
    ) : super(
        output,
        ElementRegistry.key,
        lookupProvider,
        parentProvider,
        Function { element: Element? -> ElementRegistry.elements.getKey(element).orElseThrow() })
}
